12-offer-encoding.md

```diff
@@ -290,26 +294,32 @@ A writer of an offer:
     - MUST NOT set `refunded_payment_hash`.
 
 A reader of an offer:
-  - if `features` contains unknown _odd_ bits that are non-zero:
+  - if the offer contains any unknown TLV fields greater or equal to 128:
+    - MUST NOT respond to the offer.
+  - if `offer_features` contains unknown _odd_ bits that are non-zero:
     - MUST ignore the bit.
-  - if `features` contains unknown _even_ bits that are non-zero:
+  - if `offer_features` contains unknown _even_ bits that are non-zero:
     - MUST NOT respond to the offer.
     - SHOULD indicate the unknown bit to the user.
-  - if `description` is not set:
+  - if `offer_description` is not set:
     - MUST NOT respond to the offer.
-  - if `node_id` is not set:
-    - if `paths` is not set or does not contain at least one `blinded_path`:
+  - if `offer_node_id` is not set:
+    - if `offer_paths` is not set or does not contain at least one `blinded_path`:
       - MUST NOT respond to the offer.
     - MUST use the final `onionmsg_path` `point` in the first `blinded_path` as the `node_id`.
-  - if it uses `amount` to provide the user with a cost estimate:
+  - if it uses `offer_amount` to provide the user with a cost estimate:
     - MUST warn user if amount of actual invoice differs significantly
         from that expectation.
   - SHOULD not respond to an offer if the current time is after
-    `absolute_expiry`.
+    `offer_absolute_expiry`.
   - FIXME: more!
 
 ## Rationale
 
+The entire offer is reflected in the invoice_request, both for
+completeness (so all information will be returned in the invoice), and
+so that the offer node can be stateless.
```

Comment:
The commit description says:

> 3. Add offer_metadata blob so you can do stateless offers more easily than
>   having to insert data in a blinded path.

But here the offer reflection is stated as the reason for statelessness. Seems like reflecting fields adds more complexity as you now need to keep track of unknown fields when parsing instead of simply ignoring them.

If statelessness can be achieved via the `offer_metadata` field, that leaves "completeness" for the justification for reflecting the offer. What's the rationale for needing such completeness?

Reply:
With full reflection, you only need the offer_metadata to contain the auth cookie (e.g SHA256(my-secret || offer)).  Without that, it also needs to duplicate the unreflected fields, making it potentially much larger.

Since we reflect many fields anyway, reflecting them all is actually simpler, and minimizes the offer size.